### PR TITLE
Feature/create database layer

### PR DESCRIPTION
**Pull request by khlopkov**

### CODE REVIEW

File: Arcadia.Ask/Storage/Questions/QuestionStorage.cs [hunk 1]
```diff
@@ -0,0 +1,118 @@
...
+    using Arcadia.Ask.Storage.Exceptions;
+    using Microsoft.EntityFrameworkCore;
+    using System;
+    using System.Collections.Generic;
+    using System.Linq;
+    using System.Threading.Tasks;
+
+    public class QuestionStorage : IQuestionStorage
+    {
+        private readonly DatabaseContext _dbCtx;
```
Comment (santee) [inline Arcadia.Ask/Storage/Questions/QuestionStorage.cs]: @sponkrashin or me will commit our resharper settings (code style in our team). It discourages '_' prefix usage. You would also need to apply all the suggestions/fixes from it (including braces etc)

File: Arcadia.Ask/Models/DTO/QuestionDTO.cs [hunk 2]
```diff
@@ -0,0 +1,22 @@
...
+
+        public string Text { get; set; }
+
+        public string Author { get; set; }
+
+        public DateTimeOffset PostedAt { get; set; }
+
+        public bool IsApproved { get; set; }
+
+        public IEnumerable<Guid> Votes { get; set; }
```
Comment (santee) [inline Arcadia.Ask/Models/DTO/QuestionDTO.cs]: as DTO's can (should) be user specific, there is no need to have Votes array (if we are not willing to show guids :) ). Having number and user-specific 'DidVote' property is probably easier to use on a client, and it doesn't reveal any ids

**Comment by sponkrashin** [comment]
Please, merge from master branch and fix all of your cs files with R#.

File: Arcadia.Ask/Storage/Questions/IQuestionStorage.cs [hunk 3]
```diff
@@ -0,0 +1,22 @@
+﻿namespace Arcadia.Ask.Questions
+{
+    using Arcadia.Ask.Models.DTO;
+    using System;
+    using System.Collections.Generic;
+    using System.Threading.Tasks;
+
+    public interface IQuestionStorage
+    {
+        Task<IEnumerable<QuestionDTO>> GetQuestions();
```
[the remark this answers is not in the document]
Comment (khlopkov) [reply Arcadia.Ask/Storage/Questions/IQuestionStorage.cs]: But @sponkrashin said me, that storage/repository should accept and return some DTO and manage Entity models inside them implementation. If I understood it correctly.
I understand, that it is bad practice to return DTOs in responses, and return DTOs in data layer, so will it be correct, if Hubs will manage with some new Models, which will have Response suffix (e.g. QuestionResponse)? And if i will use this, so what I should send to IQuestionClient Response or DTO?

[the remark this answers is not in the document]
Comment (sponkrashin) [reply Arcadia.Ask/Storage/Questions/IQuestionStorage.cs]: I said that database layer should operate its own internal entities like User, Question, Vote, all business logic should operate its own internal models no matter how you call them - DTOs, Models or somehow else, and controllers should operate own ones - responses and requests models. Thus we can achieve abstraction of layers from each other.

[the remark this answers is not in the document]
Comment (sponkrashin) [reply Arcadia.Ask/Storage/Questions/IQuestionStorage.cs]: In such a simple projects, I suppose, it is appropriate to use only two abstractions - one for all internal layers and one for controllers. In this way it will be like @santee wants - all logic is done over database entities and controllers return DTOs.

[the remark this answers is not in the document]
Comment (santee) [reply Arcadia.Ask/Storage/Questions/IQuestionStorage.cs]: @sponkrashin right.
In perfect world, you'd always have subsystem models encapsulated, so changes in one place do not break the other. Classes that are used to communicate between these subsystems are called DTOs. In a simple case, you'd just have everything external (requests/responses/database-entities) separated - basically, three three tier app. Hub/controller (requests/responses) <> application (domain model) <> database (EF entities). That would be alright. However, in this case it is really very simple and you do not need to distinguish database model and domain entity, as you use code-first approach and database should not be changed from outside. Compile time checks will take care about the rest.

To summarize - just use DTO's in hubs, and EF model everywhere else (storage interface).

**Comment by sponkrashin** [comment]
@khlopkov, please merge with master once again and apply Resharper and tslint to your code.

File: Arcadia.Ask/ClientApp/src/app/questions/questions-store.service.ts [hunk 4]
```diff
@@ -1,11 +1,14 @@
 import { Injectable, OnDestroy } from '@angular/core';
-import { Question } from './question';
+import { Question, QuestionImpl } from './question';
 import { HubConnectionBuilder, HubConnection } from '@aspnet/signalr';
 import { Observable, from, Subscription, ReplaySubject, ConnectableObservable } from 'rxjs';
 import { flatMap, scan, switchMap, startWith, multicast } from 'rxjs/operators';
 import { Map } from 'immutable';
 
-type QuestionChange = { type: 'added', question: Question } | { type: 'removed', id: string };
+type QuestionChange = { type: 'added', question: Question } |
```
Comment (sponkrashin) [inline Arcadia.Ask/ClientApp/src/app/questions/questions-store.service.ts]: Please, move part after equals sign to the next line. It would look nicer, I think.

File: Arcadia.Ask/ClientApp/src/app/questions/questions-store.service.ts [hunk 5]
```diff
@@ -115,16 +128,49 @@ export class QuestionsStore implements OnDestroy {
...
+        );
+        updateAcc = acc.set(change.id, questionWithUpdatedVotes);
+        break;
+
+      case 'approved':
+        const unapprovedQuestion = acc.get(change.id);
+        const approvedQuestion = new QuestionImpl(
+          unapprovedQuestion.text,
+          unapprovedQuestion.author,
+          unapprovedQuestion.votes,
```
Comment (sponkrashin) [inline Arcadia.Ask/ClientApp/src/app/questions/questions-store.service.ts]: Shouldn't it increase by 1?

File: Arcadia.Ask/ClientApp/src/app/questions/questions-store.service.ts [hunk 6]
```diff
@@ -102,7 +105,17 @@ export class QuestionsStore implements OnDestroy {
...
+      const onQuestionVotesChanged = (id: string, votes: number) => {
+        subscriber.next({ type: 'votesChanged', id, votes });
+      };
+
+      const onQuestionApproved = (id: string) => {
+        subscriber.next({ type: 'approved', id });
+      };
+
       hubConnection.on('QuestionIsChanged', onQuestionChanged);
+      hubConnection.on('QuestionVotesAreChanged', onQuestionVotesChanged);
```
Comment (santee) [inline Arcadia.Ask/ClientApp/src/app/questions/questions-store.service.ts]: why need these two?

File: Arcadia.Ask/Storage/DatabaseContext.cs [hunk 7]
```diff
@@ -0,0 +1,34 @@
...
+        }
+
+        public DbSet<QuestionEntity> Questions { get; set; }
+
+        public DbSet<VoteEntity> Votes { get; set; }
+
+        protected override void OnModelCreating(ModelBuilder modelBuilder)
+        {
+            modelBuilder.Entity<VoteEntity>()
+                .HasKey(v => new {v.QuestionId, v.UserId});
```
Comment (sponkrashin) [inline Arcadia.Ask/Storage/DatabaseContext.cs]: Have you already applied Resharper code cleanup to this file?

File: Arcadia.Ask/Hubs/QuestionsHub.cs [hunk 8]
```diff
@@ -1,31 +1,44 @@
...
     using Microsoft.AspNetCore.SignalR;
 
+    using Models.DTO;
+    using Models.Entities;
+
+    using Storage.Questions;
+
     public class QuestionsHub : Hub<IQuestionsClient>
     {
+        private readonly Guid currentUserGuid;
```
Comment (santee) [inline Arcadia.Ask/Hubs/QuestionsHub.cs]: should not be a private field. Hub lifecycle is transient (every hub method call is executed on a new hub instance), so you should not store state in a class member

File: Arcadia.Ask/Storage/Questions/QuestionStorage.cs [hunk 9]
```diff
@@ -0,0 +1,127 @@
...
+    using System;
+    using System.Collections.Generic;
+    using System.Linq;
+    using System.Threading.Tasks;
+
+    using Exceptions;
+
+    using Microsoft.EntityFrameworkCore;
+
+    using Models.Entities;
```
Comment (sponkrashin) [inline Arcadia.Ask/Storage/Questions/QuestionStorage.cs]: Both this and `using Exceptions` should have `Arcadia.Ask` part in them to be grouped together after cleanup. I set it in Resharper configuration, but it seems to be missing after some refactorings. I will fix it.

Comment (sponkrashin) [reply Arcadia.Ask/Storage/Questions/QuestionStorage.cs]: No, I've just checked and realized that the latest Resharper configuration adds `Arcadia.Ask` part to usings if it is missed. So, you should make cleanup again, I suppose.

File: Arcadia.Ask/ClientApp/src/app/questions/questions-store.service.ts (hunk 6 above)
Comment (khlopkov) [reply Arcadia.Ask/ClientApp/src/app/questions/questions-store.service.ts]: Because DTO is user specific, so when I notify user about changing votes, if send dto, then it will pass property didVote, which is user specific too. That's why I decided to add several connections for different cases. So, when someone voted, all users will get updated number of votes, and only user, who voted will get full question payload.

File: Arcadia.Ask/Storage/Questions/QuestionStorage.cs (hunk 9 above)
Comment (khlopkov) [reply Arcadia.Ask/Storage/Questions/QuestionStorage.cs]: I've just tried to do this, and it still removes Arcadia.Ask prefixes after clean up. 
Probably this is because I do it from my home PC with Rider IDE, but it includes Resharper in it, so I don't think that it is cause of problem. Also I made clean up of this file already from PC at Arcadia, with latest pulled from master configuration

Comment (sponkrashin) [reply Arcadia.Ask/Storage/Questions/QuestionStorage.cs]: Ok, we'll figure it out when you are at the office.

File: Arcadia.Ask/Storage/Questions/QuestionStorage.cs [hunk 10]
```diff
@@ -0,0 +1,127 @@
...
+            this.dbCtx.Remove(voteEntity);
+            await this.dbCtx.SaveChangesAsync();
+
+            return await this.FindQuestionEntityByIdAsync(questionId);
+        }
+
+        private async Task<QuestionEntity> FindQuestionEntityByIdAsync(Guid questionId)
+        {
+            return await this.dbCtx.Questions
+                .Where(q => q.QuestionId == questionId)
```
Comment (sponkrashin) [inline Arcadia.Ask/Storage/Questions/QuestionStorage.cs]: As you don't include `AsNoTracking` method and DB entities are used across the application as-is - it may cause hard recognizable problems, because all changes in DB entity objects you made somewhere in business logic would be tracked by EF. And any subsequent call to `SaveChanges` will save those accidental changes too.

File: Arcadia.Ask/Storage/Questions/QuestionStorage.cs [hunk 11]
```diff
@@ -0,0 +1,127 @@
...
+        public async Task<QuestionEntity> UpvoteQuestion(Guid questionId, Guid userId)
+        {
+            var entity = await this.FindQuestionEntityByIdAsync(questionId);
+
+            if (entity == null)
+            {
+                throw new QuestionNotFoundException(questionId);
+            }
+
+            if (await this.dbCtx.Votes.Where(v => v.QuestionId == questionId && v.UserId == userId).FirstOrDefaultAsync() != null)
```
Comment (sponkrashin) [inline Arcadia.Ask/Storage/Questions/QuestionStorage.cs]: Don't include so long expressions to condition blocks, it makes it difficult to read.

File: Arcadia.Ask/Models/Entities/VoteEntity.cs [hunk 12]
```diff
@@ -0,0 +1,15 @@
+﻿namespace Arcadia.Ask.Models.Entities
+{
+    using System;
+    using System.ComponentModel.DataAnnotations.Schema;
+
+    public class VoteEntity
+    {
+        [ForeignKey("Question")]
```
Comment (sponkrashin) [inline Arcadia.Ask/Models/Entities/VoteEntity.cs]: Doesn't it need to have `VoteId` also?

File: Arcadia.Ask/Startup.cs [hunk 13]
```diff
@@ -22,7 +24,12 @@ public Startup(IConfiguration configuration)
         // This method gets called by the runtime. Use this method to add services to the container.
         public void ConfigureServices(IServiceCollection services)
         {
-            services.AddSingleton<IQuestionStorage, QuestionStorage>();
+            services.AddEntityFrameworkInMemoryDatabase();
+            services.AddDbContext<DatabaseContext>((sp, options) =>
+            {
+                options.UseInMemoryDatabase().UseInternalServiceProvider(sp);
```
Comment (sponkrashin) [inline Arcadia.Ask/Startup.cs]: Method `UseInMemoryDatabase()` is obsolete.

File: Arcadia.Ask/Models/Entities/VoteEntity.cs (hunk 12 above)
Comment (khlopkov) [reply Arcadia.Ask/Models/Entities/VoteEntity.cs]: As i thought vote is unique for every user and question. So i defined complex primary key consisting of QuestionId and UserId, which described via fluentApi.
Should I add Id field for primary key and replace complex primary key with unique index?

File: Arcadia.Ask/ClientApp/src/app/questions/questions-store.service.ts (hunk 4 above)
Comment (khlopkov) [reply Arcadia.Ask/ClientApp/src/app/questions/questions-store.service.ts]: https://github.com/SoftwareCountry/arcadia-ask/pull/2/commits/1bedeb52e1f2f1419505bab22f623a9fa03d2fa9

File: Arcadia.Ask/Storage/DatabaseContext.cs (hunk 7 above)
Comment (khlopkov) [reply Arcadia.Ask/Storage/DatabaseContext.cs]: https://github.com/SoftwareCountry/arcadia-ask/pull/2/commits/c86d153c49fecceecd58cac7a49fb4f6f6482cd4

File: Arcadia.Ask/Hubs/QuestionsHub.cs (hunk 8 above)
Comment (khlopkov) [reply Arcadia.Ask/Hubs/QuestionsHub.cs]: Didn't know that. I fixed in using concurrent dictionary, mapping users guid, when hub connection initialized, and remove after disconnect.
https://github.com/SoftwareCountry/arcadia-ask/pull/2/commits/b6cbffba96c28b6d1d6fe5cce07195bb48ad7acb

File: Arcadia.Ask/Storage/Questions/QuestionStorage.cs (hunk 10 above)
Comment (khlopkov) [reply Arcadia.Ask/Storage/Questions/QuestionStorage.cs]: https://github.com/SoftwareCountry/arcadia-ask/pull/2/commits/38d0d7e409a7b9ce54e9b9200678fb4cd34dcc31

File: Arcadia.Ask/Startup.cs (hunk 13 above)
Comment (khlopkov) [reply Arcadia.Ask/Startup.cs]: https://github.com/SoftwareCountry/arcadia-ask/pull/2/commits/fc3a52b02a2d091483607960cfbf2247628c5aae

File: Arcadia.Ask/Storage/Questions/QuestionStorage.cs (hunk 11 above)
Comment (khlopkov) [reply Arcadia.Ask/Storage/Questions/QuestionStorage.cs]: https://github.com/SoftwareCountry/arcadia-ask/pull/2/commits/1537be9b7ef8a0734acf565dd58140980c6b896a

File: Arcadia.Ask/Models/Entities/VoteEntity.cs (hunk 12 above)
Comment (sponkrashin) [reply Arcadia.Ask/Models/Entities/VoteEntity.cs]: I'm not sure it will work in full functional database (not in-memory) if we change to it.

Comment (khlopkov) [reply Arcadia.Ask/Models/Entities/VoteEntity.cs]: I think it depends on database. Postgres supports complex primary keys, for example. But not sure about other databases

Comment (sponkrashin) [reply Arcadia.Ask/Models/Entities/VoteEntity.cs]: As I found in documentation, composite key defined the following way `modelBuilder.Entity<VoteEntity>().HasKey(v => new { v.QuestionId, v.UserId });` should work.

File: Arcadia.Ask/Storage/Questions/QuestionStorage.cs (hunk 9 above)
Comment (sponkrashin) [reply Arcadia.Ask/Storage/Questions/QuestionStorage.cs]: As you can see in my last commit to the current PR, resharper code cleanup has complemented namespaces up to full state. Maybe the problem is in what clenaup profile was used from your Rider IDE.

File: Arcadia.Ask/Hubs/QuestionsHub.cs [hunk 14]
```diff
@@ -36,19 +71,56 @@ public async Task RemoveQuestion(Guid questionId)
...
+        public async Task<IEnumerable<QuestionDto>> GetQuestions()
         {
-            return this.questionsStorage.GetQuestions();
+            var questions = await this.questionsStorage.GetQuestions();
+            return questions.Select(q => this.EntityToDtoForSpecificUser(q, this.CurrentUserGuid));
+        }
+
+        private async Task VotesAreChanged(QuestionDto question)
+        {
+            await this.Clients.AllExcept(this.Context.ConnectionId).QuestionVotesAreChanged(question.QuestionId, question.Votes);
```
Comment (sponkrashin) [inline Arcadia.Ask/Hubs/QuestionsHub.cs]: @khlopkov, just for curiosity - what was the main reason to do so?

File: Arcadia.Ask/Hubs/QuestionsHub.cs [hunk 15]
```diff
@@ -36,19 +71,56 @@ public async Task RemoveQuestion(Guid questionId)
...
+
+            return new QuestionDto
+            {
+                QuestionId = entity.QuestionId,
+                Author = entity.Author,
+                Text = entity.Text,
+                PostedAt = entity.PostedAt,
+                IsApproved = entity.IsApproved,
+                Votes = entity.Votes?.Count ?? 0
+            };
```
Comment (sponkrashin) [inline Arcadia.Ask/Hubs/QuestionsHub.cs]: What about `DidVote` property?

Comment (sponkrashin) [reply Arcadia.Ask/Hubs/QuestionsHub.cs]: Nevermind, I've already noticed it in another method.

File: Arcadia.Ask/Hubs/QuestionsHub.cs (hunk 14 above)
Comment (khlopkov) [reply Arcadia.Ask/Hubs/QuestionsHub.cs]: As I mentioned earlier. DTOs have user specific property 'didVote'. So when votes are changed if to all clients dto will be passed, then they have invalid didVote flag. As a solution I also can send another DTO without DidVote property, and merge passed in it fields with current  values on frontend. But I thought, that it will be better if events be more specific.

File: Arcadia.Ask/Models/Entities/VoteEntity.cs (hunk 12 above)
Comment (khlopkov) [reply Arcadia.Ask/Models/Entities/VoteEntity.cs]: Yes. I've already done it

File: Arcadia.Ask/Hubs/QuestionsHub.cs (hunk 8 above)
Comment (santee) [reply Arcadia.Ask/Hubs/QuestionsHub.cs]: still a no :) you should never use static variables to keep state. Reason is their lifecycle. You'll never be able to write tests, for example, or scale it.

Comment (santee) [reply Arcadia.Ask/Hubs/QuestionsHub.cs]: proper way to achieve what you want is to write a service with a state and inject it

File: Arcadia.Ask/Hubs/QuestionsHub.cs (hunk 14 above)
Comment (santee) [reply Arcadia.Ask/Hubs/QuestionsHub.cs]: invalid didVote flag... ok, you're right, we need to think on that

File: Arcadia.Ask/Hubs/QuestionsHub.cs (hunk 8 above)
Comment (santee) [reply Arcadia.Ask/Hubs/QuestionsHub.cs]: that is, if you want to store a state somewhere. I'm pretty sure where is a better way to assign to get a user in hubs, will investigate it tomorrow

File: Arcadia.Ask/Storage/Questions/QuestionStorage.cs (hunk 10 above)
Comment (sponkrashin) [reply Arcadia.Ask/Storage/Questions/QuestionStorage.cs]: You don't really need `FindDetachedQuestionEntityByIdAsync` method. All you need to do is to call `DetachQuestionEntity` prior to return value from storage method. You are working with attached ordinal DB entities inside storage and return detached entities to outside.

File: Arcadia.Ask/Hubs/QuestionsHub.cs (hunk 14 above)
Comment (sponkrashin) [reply Arcadia.Ask/Hubs/QuestionsHub.cs]: What about sending correct `QuestionDTO` object for each client? You have mapping between Client and user id, you can generate proper question response for each client, and you have just to call `Client.QuestionIsChanged` method for each client.

Comment (khlopkov) [reply Arcadia.Ask/Hubs/QuestionsHub.cs]: I think, that later users Guid will not be mapped on server side. Solution with mapping guids in dictionary is temporary, because it generate new Guid for connection, so if user will reload page, then it will generate new Guid. I like solution with generating guid on client's page using cookie, but probably @santee will find another solution. 
So if it happens, it is will be imposible to send to every client each `QuestionDTO`. Or it will be necessary to store mapped usersId on server too.

Comment (sponkrashin) [reply Arcadia.Ask/Hubs/QuestionsHub.cs]: It will be, because anyway you'll need to map clients to guids, no matter where they are stored. And if later we force users to sign-in before voting, we will be able to map by user name instead of GUID.

Comment (khlopkov) [reply Arcadia.Ask/Hubs/QuestionsHub.cs]: Ok. 
In this review https://github.com/SoftwareCountry/arcadia-ask/pull/2#pullrequestreview-219530858 @santee wrote, that i should not store state in static variable, and should create service, for it. So should this service registered as singletone?

Comment (sponkrashin) [reply Arcadia.Ask/Hubs/QuestionsHub.cs]: I don't insist that my suggestion is the best, it's just an alternative way to achieve the goal.
Yes, this service should be singleton, otherwise how should it persist mapping between requests?
